### ISSUE TITLE
[QUESTION]--fp4-recipe nvfp4 works but overall performance degrades for Miles/Megatron SFT

### ISSUE THREAD

**Issue by deepak-vij** [NVIDIA/Megatron-LM#3470]
**Your question**
Hi @Phlip79 , this a follow up question to my earlier issue we closed recently.

I was finally able to successfully run the SFT using Miles/Megatron frameworks for nvfp4 configuration option. Now I do see nvfp4 specific kernels being invoked too in the pytorch/profiler Perfetto traces. However, I observed that the overall performance (time duration) of the training run has degraded substantially. I was hoping the other way around, overall performance to get better. This is based on the pytorch profiler traces I had with and without nvfp4.  

I am doing SFT for post training "Qwen3-VL-8B-Instruct" model using the following cosmos video training dataset (100 samples) on B200.

It would be great to get community feedback on this. Thanks. 

https://huggingface.co/datasets/nvidia/Cosmos-Reason1-SFT-Dataset/tree/main/agibot

**Comment by Phlip79** [comment]
Can you provide a reproducer within Megatron-LM?

Also, I would recommend looking into *what* slowed down (not just e2e). Did the individual kernels get slower? Or are you now waiting on dataloaders? Or ...? You should be able to figure that out using the Perfetto traces you already generated.

**Comment by deepak-vij** [comment]
Hi @Phlip79 , I will rerun the training jobs and analyze profiler traces in detail. Also, as I am still hacking away at the Miles/Megatron integration with the latest release of Megatron, I am hardcoding the following code in extensions/transformer_engine.py. This is because I have not figured out how to incorporate --te-precision-config-file argument in Megatron yet. So essentially, I am currently running without any .yaml file. As a result, I am doing testing this hacky way. I do see the nvf8 kernels in the perfetto tracing which did not exist before. Meaning that nvfp4 seems to be working, I think.

Also, would it be possible for you to share a sample yaml file for defining --te-precision-config-file argument. That would be quite helpful for me. It seems without this yaml file, "qparams: TEQuantizationParams " are not enabled. Thanks.   

def _get_fp8_autocast_for_quant_params(qparams: TEQuantizationParams | None, training: bool):
    
    if qparams is None:
        # print("!!! FORCING NVFP4 CONTEXT (Blackwell e2m1) !!!")
        # quant_recipe = te.common.recipe.NVFP4BlockScaling()
        print("!!! FORCING MXFP8 CONTEXT with MXBlkScaling (Blackwell e4m3) !!!")
        # quant_recipe = te.common.recipe.Float8CurrentScaling(fp8_format = te.common.recipe.Format.E4M3)
        # quant_recipe = te.common.recipe.Float8BlockScaling(fp8_format = te.common.recipe.Format.E4M3)
        quant_recipe = te.common.recipe.MXFP8BlockScaling(fp8_format = te.common.recipe.Format.E4M3)
        return te.pytorch.fp8_autocast(enabled=True, fp8_recipe=quant_recipe)
    # ------------------------------

**Comment by deepak-vij** [comment]
@Phlip79 , in continuation to my previous message, here are the profiler numbers I just ran again. Thanks.

**Without nvfp4:**

| Category          | Time Duration (us) | Percentage (%) |
| ----------------- | ------------------ | -------------- |
| Average Step Time | 4,681,112          | 100            |
| Kernel            | 2,318,990          | 49.54          |
| Memcpy            | 169,372            | 3.62           |
| Memset            | 2,434              | 0.05           |
| Communication     | 1,298              | 0.03           |
| Runtime           | 0                  | 0              |
| DataLoader        | 0                  | 0              |
| CPU Exec          | 1,439,892          | 30.76          |
| Other             | 749,127            | 16             | 

**With nvfp4**:

| Category          | Time Duration (us) | Percentage (%) |
| ----------------- | ------------------ | -------------- |
| Average Step Time | 7,645,981          | 100            |
| Kernel            | 4,777,327          | 62.48          |
| Memcpy            | 168,742            | 2.21           |
| Memset            | 2,241              | 0.03           |
| Communication     | 1,451              | 0.02           |
| Runtime           | 0                  | 0              |
| DataLoader        | 0                  | 0              |
| CPU Exec          | 1,826,564          | 23.89          |
| Other             | 869,656            | 11.37          |

**With mxfp8**:

| Category          | Time Duration (us) | Percentage (%) |
| ----------------- | ------------------ | -------------- |
| Average Step Time | 6,365,782          | 100            |
| Kernel            | 2,783,947          | 43.73          |
| Memcpy            | 167,608            | 2.63           |
| Memset            | 2,249              | 0.04           |
| Communication     | 1,887              | 0.03           |
| Runtime           | 0                  | 0              |
| DataLoader        | 0                  | 0              |
| CPU Exec          | 2,360,232          | 37.08          |
| Other             | 1,049,859          | 16.49          |

**Comment by euronymous-aithal** [comment]
+ @guyueh1  for Viz

**Comment by Phlip79** [comment]
A few questions:

- Is `Kernel` time above host or device?
- "Without nvfp4": does this mean bf16?
- Can you include the perfetto traces?
  - Please ensure print statements are removed when gathering these
- Have you tried using the [fp4](https://github.com/NVIDIA/Megatron-LM/blob/0d0943c6bfa9cbb30fcd62d40ce1792c4cb201e8/megatron/core/transformer/transformer_config.py#L519) pr [fp8](https://github.com/NVIDIA/Megatron-LM/blob/0d0943c6bfa9cbb30fcd62d40ce1792c4cb201e8/megatron/core/transformer/transformer_config.py#L438) flags?
  - I have never used Miles before so I'm not sure if this is a reasonable question or not
  - [Here](https://github.com/NVIDIA/Megatron-LM/blob/0d0943c6bfa9cbb30fcd62d40ce1792c4cb201e8/examples/rl/model_configs/nemotron5_56b.sh#L63-L70) is an example of how fp8 is used
- Can you please share the Megatron configs you are using?

**Comment by deepak-vij** [comment]
@Phlip79 
> A few questions:
> 
> * Is `Kernel` time above host or device?
> * "Without nvfp4": does this mean bf16?
> * Can you include the perfetto traces?
>   
>   * Please ensure print statements are removed when gathering these
> * Have you tried using the [fp4](https://github.com/NVIDIA/Megatron-LM/blob/0d0943c6bfa9cbb30fcd62d40ce1792c4cb201e8/megatron/core/transformer/transformer_config.py#L519) pr [fp8](https://github.com/NVIDIA/Megatron-LM/blob/0d0943c6bfa9cbb30fcd62d40ce1792c4cb201e8/megatron/core/transformer/transformer_config.py#L438) flags?
>   
>   * I have never used Miles before so I'm not sure if this is a reasonable question or not
>   * [Here](https://github.com/NVIDIA/Megatron-LM/blob/0d0943c6bfa9cbb30fcd62d40ce1792c4cb201e8/examples/rl/model_configs/nemotron5_56b.sh#L63-L70) is an example of how fp8 is used
> * Can you please share the Megatron configs you are using?

1) Kernel is as it shows up in pytorch profiler. All these numbers are from the profiler. Kernel is GPU-kernel as the profiler information shows up in Tensorboard, if this is what you meant. I just copied the numbers from Tensorboard report.
2) Yes, I am doing post training (SFT) on top of "Qwen/Qwen3-VL-8B-Instruct", which is bf16.
3) My configs are as below.
4) Regarding sending profiler traces, these are large files. There might be security issues with uploading these large file. Is it possible to do a quick Zoom call, I can show and describe all the details. I am located in the bay area (San Jose). Please let me know if it is convenient, I can send out a Zoom invite accordingly. Thanks. 

SFT_ARGS=(
   --rollout-function-path miles.rollout.sft_rollout.generate_rollout
   --prompt-data /root/datasets/${DATASET_LOCAL_NAME}/miles_sft_final_100.jsonl
   --input-key messages
   --rollout-shuffle
   --num-epoch 100
   --rollout-batch-size 96
   --transformer-impl transformer_engine
    --fp4-format 'e2m1'
    --fp4-recipe nvfp4
    --fp4-param-gather
   --num-steps-per-rollout 1
   --loss-type sft_loss
   --calculate-per-token-loss
   --disable-compute-advantages-and-returns
   --debug-train-only
   --tensor-model-parallel-size 4
   
BACKEND_ARGS=(
      --train-backend megatron
      --tensor-model-parallel-size 4
      --sequence-parallel
      --pipeline-model-parallel-size 1
      --context-parallel-size 1
      --expert-model-parallel-size 1
      --expert-tensor-parallel-size 1
      --recompute-method uniform
      --recompute-num-layers 1
      --use-dynamic-batch-size
      --max-tokens-per-gpu 4096
      --attention-dropout 0.0
      --hidden-dropout 0.0
      --accumulate-allreduce-grads-in-fp32
      --attention-softmax-in-fp32
      --attention-backend flash
      --megatron-to-hf-mode bridge
      --use-pytorch-profiler
      --tensorboard-dir ./profiler_logs
      --profile-step-start 30
      --profile-step-end 32
    )

**Comment by Phlip79** [comment]
From @erhoo82:

mxfp8 and nvfp4 being slower is expected when the workload uses small tensors. mxfp8 and nvfp4 calls extra kernels for online quantization, and will slow down.

There are two solutions.

1. Use large per-GPU tensors by increasing sequence packing for SFT
2. Wait for the on-going optimization to fuse the casting operations

2 can help the perf a bit, but eventually you will need 1 to make these low precision recipe really shine.

**Comment by deepak-vij** [comment]
Thanks @Phlip79 for your response. I am a bit confused now. Are you implying that even if we do option 1) & 2), we may not see   performance benefits during training? In that case, are we better of not using fp4/fp8 precisions at all at this time as currently training performance seems to degrade? Is there something else need to happen before we start seeing value proposition of low precision fp4/fp8 datatypes processing during training?  

Regarding option 1, increasing sequence packing at the input dataset level for SFT. If I am not mistaken, sequence packing has implications on the cudagraph triggering implications as well. Please see other issue I created earlier regarding cudagraph triggering as part of SFT training. https://github.com/NVIDIA/Megatron-LM/issues/3450#issuecomment-3918039886 

For option 2, is that something we can collaborate in order to expedite the effort, if makes sense.

Also, not clear on your last comment, making low precision recipe to shine. Thanks again.  
>> but eventually they need 1 to make these low precision recipe really shine.

**Comment by Phlip79** [comment]
Sorry for not being clear: you will want to do option 1 no matter what (even once option 2 is available). 4 and 8 bit math is extremely fast; however, it requires online quantization.

- small tensors: online quantization can dominate total kernel time
- large tensors: online quantization becomes negligible and you can truly take advantage of 4/8 bit TFLOPs

The larger the sequence, the larger the tensor: thus a larger sequence will be able to take advantage of 4/8 bit TFLOPs. Done correctly, you will see large performance benefits.

**Comment by deepak-vij** [comment]
Thanks @Phlip79, really appreciate back&forth conversation on this very important topic. I am definitely going to try out the suggestion you described. Also, it would be great if you can share any existing example codebase for increasing sequence packing for SFT.

Finally, how prevalent is this approach. Has industry started adopting such full-quantized SFT training approach in their actual production environment? You folks at megatron included this very recently too as part of the open source codebase. 

I could definitely see the benefits of this approach for use cases such as autonomous driving/robotics where it is desirable to keep the master weights at higher precision (bf16) but do post training using fp4/fp8. I did see Nvidia paper on this as well.

https://arxiv.org/pdf/2505.19115

**Comment by Phlip79** [comment]
> Also, it would be great if you can share any existing example codebase for increasing sequence packing for SFT.

Megatron-Bridge has a [good doc](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/docs/training/packed-sequences.md) about this. If you're unfamiliar with Megatron-Bridge: think of it as a Megatron-Core "extension" bridging HF models and Megatron-Core. Packed sequences was actually recently [enabled by default](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/2284) for all Megatron-Bridge recipes.

> Has industry started adopting such full-quantized SFT training approach in their actual production environment?

Definitely.

**Comment by deepak-vij** [comment]
Hi @Phlip79 , would you have any example case study of who is using this approach to do Post-training quantization (compute time). Thanks.

**Comment by Phlip79** [comment]
We will have more to share over the next few months, but here is what I found that is public:

- [FP8 Quantization in NeMo RL](https://docs.nvidia.com/nemo/rl/latest/fp8.html)
- [LMSYS FP8 during RL](https://lmsys.org/blog/2025-11-25-fp8-rl)
  - Looks like the Miles team helped with this
- [FP8 RL Research Paper](https://www.arxiv.org/pdf/2601.18150)
- [Another FP8 RL Research Paper](https://arxiv.org/html/2601.14243v2)
  - Includes NVIDIA authors
- [unsloth doc](https://unsloth.ai/docs/get-started/reinforcement-learning-rl-guide/fp8-reinforcement-learning)

**Comment by deepak-vij** [comment]
Thanks @Phlip79.

**Comment by Phlip79** [comment]
Feel free to re-open this issue or open a new issue with any other questions
